sdk/storage/Azure.Storage.DataMovement/src/JobPartInternal.cs

```diff
@@ -0,0 +1,209 @@
+            CancellationTokenSource cancellationTokenSource)
+        {
+            JobPartStatus = StorageTransferStatus.Queued;
+            PartNumber = partNumber;
+            _dataTransfer = dataTransfer;
+            _sourceResource = sourceResource;
+            _destinationResource = destinationResource;
+            _errorHandling = errorHandling;
+            _createMode = createMode;
+            _checkpointer = checkpointer;
+            _cancellationTokenSource = cancellationTokenSource;
+            _maximumTransferChunkSize = maximumTransferChunkSize;
+            _initialTransferSize = initialTransferSize;
+            _arrayPool = arrayPool;
+            TransferStatusEventHandler = statusEventHandler;
+            TransferFailedEventHandler = failedEventHandler;
+        }
+
+        public void SetQueueChunkDelegate(QueueChunkDelegate chunkDelegate)
+        {
+            QueueChunk = chunkDelegate;
+        }
+
+        /// <summary>
+        /// Processes the job to job parts
+        /// </summary>
+        /// <returns>An IEnumerable that contains the job chunks</returns>
+        public abstract Task ProcessPartToChunkAsync();
+
+        internal async Task TriggerCancellation()
+        {
+            if (!_cancellationTokenSource.IsCancellationRequested)
+            {
+                _cancellationTokenSource.Cancel();
+            }
+            await OnTransferStatusChanged(StorageTransferStatus.Completed).ConfigureAwait(false);
+            _dataTransfer._state.ResetTransferredBytes();
+        }
+
+        /// <summary>
+        /// To change all transfer statues at the same time
+        /// </summary>
+        /// <param name="transferStatus"></param>
+        internal async Task OnTransferStatusChanged(StorageTransferStatus transferStatus)
+        {
+            if (JobPartStatus != transferStatus)
+            {
+                JobPartStatus = transferStatus;
+                // TODO: change to RaiseAsync
+                if (TransferStatusEventHandler != null)
+                {
+                    await TransferStatusEventHandler.Invoke(new TransferStatusEventArgs(
+                        _dataTransfer.Id,
+                        transferStatus,
+                        false,
+                        _cancellationTokenSource.Token)).ConfigureAwait(false);
+                }
+            }
+        }
+
+        /// <summary>
+        /// To change all transfer statues at the same time
+        /// </summary>
+        /// <param name="bytesTransferred"></param>
+        internal void ReportBytesWritten(long bytesTransferred)
+        {
+            _dataTransfer._state.UpdateTransferBytes(bytesTransferred);
+        }
+
+        /// <summary>
+        /// Invokes Failed Argument
+        /// </summary>
+        internal async Task InvokeFailedArg(Exception ex)
+        {
+            if (TransferFailedEventHandler != null)
+            {
+                // TODO: change to RaiseAsync
```

Comment:
TODO

Reply:
https://github.com/Azure/azure-sdk-for-net/issues/32734